invokeai/frontend/web/src/app/NodeAPITest.tsx

```diff
@@ -0,0 +1,200 @@
+import { Flex, Image, Text } from '@chakra-ui/react';
+import IAIButton from 'common/components/IAIButton';
+import {
+  setProgress,
+  setProgressImage,
+  setSessionId,
+  setStatus,
+} from 'services/apiSlice';
+import { useEffect } from 'react';
+import { STATUS, ProgressImage } from 'services/apiSliceTypes';
+import { getImage } from 'services/thunks/image';
+import {
+  cancelProcessing,
+  createSession,
+  invokeSession,
+} from 'services/thunks/session';
+import { io } from 'socket.io-client';
+import { useAppDispatch, useAppSelector } from './storeHooks';
+import { RootState } from './store';
+
+type GeneratorProgress = {
+  session_id: string;
+  invocation_id: string;
+  progress_image: ProgressImage;
+  step: number;
+  total_steps: number;
+};
+
+const socket_url = `ws://${window.location.host}`;
+const socket = io(socket_url, {
+  path: '/ws/socket.io',
+});
+
+const NodeAPITest = () => {
+  const dispatch = useAppDispatch();
+  const { sessionId, status, progress, progressImage } = useAppSelector(
+    (state: RootState) => state.api
+  );
+
+  const handleCreateSession = () => {
+    dispatch(
+      createSession({
+        requestBody: {
+          nodes: {
+            a: {
+              id: 'a',
+              type: 'txt2img',
+              prompt: 'pizza',
+              steps: 50,
+              seed: 123,
+            },
+            b: {
+              id: 'b',
+              type: 'img2img',
+              prompt: 'dog',
+              steps: 50,
+              seed: 123,
+              strength: 0.9,
+            },
+            c: {
+              id: 'c',
+              type: 'img2img',
+              prompt: 'cat',
+              steps: 50,
+              seed: 123,
+              strength: 0.9,
+            },
+          },
+          edges: [
+            {
+              source: { node_id: 'a', field: 'image' },
+              destination: { node_id: 'b', field: 'image' },
+            },
+            {
+              source: { node_id: 'b', field: 'image' },
+              destination: { node_id: 'c', field: 'image' },
+            },
+          ],
+        },
+      })
+    );
+  };
+
+  const handleInvokeSession = () => {
+    dispatch(invokeSession());
+  };
+
+  const handleCancelProcessing = () => {
+    dispatch(cancelProcessing());
+  };
+
+  useEffect(() => {
+    if (!sessionId) {
+      return;
+    }
+
+    // set up socket.io listeners
+
+    // TODO: suppose this should be handled in the socket.io middleware
+    // TODO: write types for the socket.io payloads, haven't found a generator for them yet...
+
+    // subscribe to the current session
+    socket.emit('subscribe', { session: sessionId });
```

Comment:
Where is the listener for this on the backend?

Reply:
So each time we hit the REST API to create a session, we will also initialize this socket subscriber for each one? And we have a separate connection (room? still learning the socket lingo) for the global state like server connection? Is that why there are two socket proxies in the vite build file now?

Reply:
> Where is the listener for this on the backend?

Listener is here: https://github.com/psychedelicious/InvokeAI/blob/f02be7ed34798307a8000b4ad9ecc75841f64a32/invokeai/app/api/sockets.py

Reply:
> So each time we hit the REST API to create a session, we will also initialize this socket subscriber for each one? And we have a separate connection (room? still learning the socket lingo) for the global state like server connection? Is that why there are two socket proxies in the vite build file now?

This is probably the wrong approach in the long term, but yes in the little demo the client is entering a new room, generating, then leaving when finished generating.

There is still a client connection that is established during that initial HTTP request to the socketio server. socketio identifies connectio with a client `sid` - session id - no relation to nodes "sessions", except that the node session id is exactly that `sid` (😅).

In the dev tools, we are looking at that overall socketio session, and then within that we may subscribe to any number of rooms.

The two proxies in the vite config proxy requests from the vite dev server (on port 5173) to the actual socketio server. One of the proxy setups is for the old server, and the second proxy is for the nodes server. 

The proxy is for the overall socketio connection; the room subscriptions, progress events, etc are all happening within that connection